### DISCUSSION TITLE
Create git-sync deployment with persistence enabled in Helm Chart

### DISCUSSION THREAD

**Discussion by j1wonpark** [apache/airflow#34984] [Q&A] [answered]
I am a devops engineer who operates several airflows based on Kubernetes executor. Analysts using airflow upload hundreds of MB of models to a git repo, and it is difficult for me to control them all. I don't want the executor to download hundreds of MB of new files every time it creates a task execution pod.

I read this [issue](https://github.com/apache/airflow/discussions/32146).

Obviously git-sync does not support parallelism, and the sidecar pattern of git-sync with multiple schedulers and running on each scheduler potentially causes a lot of problems.

So, in the case of a scheduler used with persistence, how about performing git-sync at the deployment (number of replicas is 1) level? If you perform git-sync in a separate pod instead of mounting git-sync for each container inside the pod, and only read in the scheduler, the potential issues mentioned above will be resolved.

**Comment by potiuk** [comment]
The problem IMHO is with storing the models  in your Git. If you keep on adding those binary artifacts to git, your repository will get enormous really quickly (your git repo stores all the history of all your files in all the repository clonse you have - this means that checking out your repository will downloads gigabytes or even terabates of your historical binary artiffacts)

Git is not well suited to store 100s of MB of models. Git should generally be used to store source files, not binary artifacts.  The best you should do it is to work out a pattern that your models are stored elsewhere (S3, GCS are the obvious candidates) and where your tasks are using those as binary model sources for the model execution - and the sources should contain references to these. You can control it on various levels in your git - for example using pre-commit or pull request review - you can even automate checks that your analysts are not commiting binary files to git.

You can also (if you really want to use git) use git-lfs which largely automates it away (https://github.com/git-lfs/git-lfs). Git lfs is the extension to git that allows you to store binary files on an external storage The way it works it that it stores references to the binary artifacts in git and it will automatically pull only the actual artifact your commit usses when needed - so basically automating the pattern with s3/GCS I described above

**Reply by j1wonpark** [reply]
Thank you for good information! However, as mentioned above, we cannot completely restrict whether customers upload models or metadata that is quite large in size. I provide a self-developed managed airflow service rather than a custom airflow for each customer.

Return to the subject, if you use git sync with pvc (a combination that is currently allowed in helm charts), what about the idea of separating gitsync into a separate deployment, keeping only 1 and attaching it to the shared volume? (not sidecar container)

**Reply by potiuk** [reply, marked as answer]
> We cannot completely restrict whether customers upload models or metadata that is quite large in size

I am not the one to say of course but if you cant, it might kill your business possibly some day if you will not solve it (but of course you have to analyse and decide). This is a free advice you get from a random maintainer here and you are responsible to make the right decisions. BTW.  If you are note sure about all the technicalities there and you need to run your business based on it, If I were you, I would likely hire and pay some expert who will look at it and will give you specific advise based on your case because IMHO you are putting yourself in danger. But of course, it might be completely wrong assessment based on a few paragraphs of text - as much free advice you can get.

You will have the same problem with PVCs keeping and distributing those models and huge `.git` repo together. That will be also multiplied by huge factor because git sync works in specific way - I warned long time ago against doing exactly what you propose in https://medium.com/apache-airflow/shared-volumes-in-airflow-the-good-the-bad-and-the-ugly-22e9f681afca and you can read all the details there. Read about it, think, apply to your case and decide. 

Depending on characteristics of your PVC it will either kill your deployment now or in some time in various ways.

There is no magic, PVC is just a networked drive that has to distribute the files anyway and the way git sync works (See article) makes it particularly vulnerable - and very much dependable on how PVC handles symbolic link and the fact that `git-sync` uses it to achieve atomic updates. 

But if you want to try it and see if it is better for you, feel free, you can always try it and see, maybe your PVC will handle it better, who knows. 

IMHO you will be shooting yourself in the foot by having huge blobs in git repo. It will kill your dreployment some day (sooner than you think) if you allow that. 

Your aproach is simply broken by using hammer as a screwdriver basically (i.e. storing huge blobs in git repo).

**Reply by j1wonpark** [reply]
Thank you for your comment. It helped me a lot in decision making!
